apps/src/codebridge/FileBrowser/dragAndDropUtils.ts

```diff
@@ -0,0 +1,188 @@
+export const fileBrowserKeyboardCoordinateGetter: (
+  folders: Record<string, ProjectFolder>
+) => KeyboardCoordinateGetter = folders => (event, args) => {
+  if (event.code !== KeyboardCode.Up && event.code !== KeyboardCode.Down) {
+    return;
+  }
+  event.preventDefault();
+
+  const {context} = args;
+  const {droppableRects, over, active} = context;
+  let orderedRects = Array.from(droppableRects.keys());
+  const currentId = active?.data.current?.id;
+  const dragType = active?.data.current?.type;
+
+  if (dragType === DragType.FOLDER) {
+    // We can't move a folder into a folder that is a child of itself
+    const allChildren = getFolderChildren(
+      currentId as string,
+      Object.values(folders)
+    );
+    orderedRects = orderedRects.filter(
+      id => !allChildren.includes(id as string)
+    );
+  }
+
+  // Sort the available droppable rectangles by their top coordinate, except
+  // for the root folder (DEFAULT_FOLDER_ID), which should always be last, as it
+  // is the folder that contains all the other folders.
+  orderedRects.sort((a, b) => {
+    // Default folder is always last in the list.
+    if (a === DEFAULT_FOLDER_ID) {
+      return 1;
+    } else if (b === DEFAULT_FOLDER_ID) {
+      return -1;
+    }
+    // Otherwise, sort based on the top coordinate.
+    const aTop = droppableRects.get(a)?.top || 0;
+    const bTop = droppableRects.get(b)?.top || 0;
+    return aTop - bTop;
+  });
+  const currentIndex = orderedRects.indexOf(over?.id as string);
+  let nextIndex = currentIndex;
+  // Find the next index based on the key pressed.
+  if (event.code === KeyboardCode.Down) {
+    nextIndex = Math.min(currentIndex + 1, orderedRects.length - 1);
+  } else if (event.code === KeyboardCode.Up) {
+    nextIndex = Math.max(currentIndex - 1, 0);
+  }
+
+  if (nextIndex === currentIndex) {
+    // No need to move if we are not changing the index (we are likely at the
+    // top or bottom of the browser).
+    return;
+  }
+  const newRectId = orderedRects[nextIndex];
+  const newRect = droppableRects.get(orderedRects[nextIndex]);
+  if (newRect) {
+    const x = newRect.left;
+    // Normally, the y coordinate is the top of the droppable area, plus an offset.
+    let y = newRect.top + FOLDER_DROP_OFFSET;
+    // If we are dropping into the root folder, we want to drop at the bottom of the
+    // droppable area, because the root folder is the entire file browser. We subtract the
+    // offset to be a bit above the bottom of the droppable area.
+    if (newRectId === DEFAULT_FOLDER_ID) {
+      y = newRect.bottom - FOLDER_DROP_OFFSET;
+    }
+    const newCoordinates = {
+      x,
+      y,
+    };
+
+    return newCoordinates;
+  }
+};
+
+// Custom collision detection algorithm for the file browser.
+// We want to drop into the folder that the file is below the top of, but if
+// we are below multiple folders, we will pick the lower of the two (which is the one
```

Comment:
Should this be "...if we are below multiple folders, we will pick the *lowest of them*..."?